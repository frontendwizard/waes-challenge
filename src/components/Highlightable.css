.wrapper {
  position: relative;
  padding: 32px;
}
.text {
  height: 250px;
  width: 100%;
  text-align: left;
  font-size: 16px;
  line-height: 1.2;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen,
    Ubuntu, Cantarell, "Open Sans", "Helvetica Neue", sans-serif;
  z-index: -1;
  overflow: scroll;
  white-space: pre-wrap;
  overflow-wrap: break-word;
  border: none;
  background: transparent;
}
.highlight {
  position: absolute;
  color: transparent;
  padding: 2px;
  width: calc(100% - 64px);
  transform: translateY(calc(-100% - 4px));
  flex-direction: column;
}
