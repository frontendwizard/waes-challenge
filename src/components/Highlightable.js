import React, { useRef } from "react"
import "./Highlightable.css"

const Highlightable = ({
  highlights,
  addHighlight,
  removeHighlight,
  color,
  text,
  onTextChange,
}) => {
  const textRef = useRef(null)
  const highlightsRef = useRef(null)

  const onScroll = e => {
    e.persist()
    highlightsRef.current.scrollTop = e.target.scrollTop
  }

  const onMouseUp = e => {
    e.persist()
    const start = textRef.current.selectionStart
    const end = textRef.current.selectionEnd
    // no selection made
    if (start === end) return false
    // check if overlaps any existing hightlight
    if (highlights.find(hg => hg.range.start === start && hg.range.end === end))
      return false
    highlights.forEach(highlight => {
      if (
        // ends in the middle of a selection
        (end > highlight.range.start && end <= highlight.range.end) ||
        // starts in the middle of a selection
        (start >= highlight.range.start && start < highlight.range.end) ||
        // selection is in the middle of the new selection
        (start <= highlight.range.start && end > highlight.range.end)
      )
        removeHighlight(highlight.id)
    })
    addHighlight({
      range: { start, end },
      color,
    })
  }

  const parseHighlights = () => {
    let extraMarkupChars = 0
    return highlights.reduce((acc, selection) => {
      const highlight =
        `${acc.slice(0, selection.range.start + extraMarkupChars)}` +
        `<span class="${selection.color}" data-testid="highlight">${acc.slice(
          selection.range.start + extraMarkupChars,
          selection.range.end + extraMarkupChars,
        )}</span>` +
        `${acc.slice(selection.range.end + extraMarkupChars)}`
      // `<span class="`. length
      // + color.length
      // + `" data-testid="highlight">`.lenth
      // + `</span>`.length
      extraMarkupChars += 13 + selection.color.length + 26 + 7
      return highlight
    }, text)
  }
  return (
    <div className="wrapper">
      <textarea
        data-testid="highlightable-textarea"
        className="text"
        ref={textRef}
        value={text}
        onChange={e => onTextChange(e.target.value)}
        onMouseUp={onMouseUp}
        onScroll={onScroll}
      />
      <div
        data-testid="highlightable-background"
        ref={highlightsRef}
        className="highlight text"
        dangerouslySetInnerHTML={{
          __html: parseHighlights(text),
        }}
      />
    </div>
  )
}

export default Highlightable
