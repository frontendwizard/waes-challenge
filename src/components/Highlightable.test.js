import React from "react"
import { render } from "react-testing-library"
import Highlightable from "./Highlightable"

describe("Highlightable", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Highlightable
        highlights={[]}
        addHighlight={() => {}}
        removeHighlight={() => {}}
        nextId={0}
        color="red"
        text="this is a test"
        onTextChange={() => {}}
      />,
    )

    expect(asFragment()).toMatchSnapshot()
  })
  it("should highlight the provided ranges", () => {
    const { getAllByTestId } = render(
      <Highlightable
        highlights={[
          { range: { start: 0, end: 4 }, color: "red", id: 0 },
          { range: { start: 5, end: 7 }, color: "blue", id: 1 },
        ]}
        addHighlight={() => {}}
        removeHighlight={() => {}}
        nextId={2}
        color="red"
        text="this is a test"
        onTextChange={() => {}}
      />,
    )

    const highlights = getAllByTestId("highlight")
    expect(highlights).toBeTruthy()
    expect(highlights).toHaveLength(2)
    expect(highlights[0].textContent).toBe("this")
    expect(highlights[1].textContent).toBe("is")
  })
})
