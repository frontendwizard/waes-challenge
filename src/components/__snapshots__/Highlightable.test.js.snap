// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlightable should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <textarea
      class="text"
      data-testid="highlightable-textarea"
    >
      this is a test
    </textarea>
    <div
      class="highlight text"
      data-testid="highlightable-background"
    >
      this is a test
    </div>
  </div>
</DocumentFragment>
`;
